### PR TITLE
Add layer control tree plugin

### PR DESCRIPTION
**Pull request by hansthen**
Implemented the `Leaflet.Control.Layers.Tree` in Folium. See https://github.com/jjimenezshaw/Leaflet.Control.Layers.Tree for more details.

Closes: #1606

### CODE REVIEW

File: tests/test_template.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+                "children": [
+                    {"label": "trail", "layer": trail},
+                    {"jscode": JsCode('function(){return "hi"}')},
+                ],
+            }
+        ],
+    }
+    js = tojavascript(d)
+    assert "poly_line" in js
+    assert 'return "hi"' in js
```
Comment (Conengmo) [inline tests/test_template.py:30]: Suggestion: maybe validate the full output here?

File: folium/template.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    )
+
+
+class Environment(jinja2.Environment):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.filters["tojavascript"] = tojavascript
+
+
+class Template(jinja2.Template):
```
Comment (Conengmo) [inline folium/template.py:50]: This PR made me realize we can adopt this new Template piecemeal, don't have to do a big bang.

I do want to merge this new module separately, so we don't sneak it in in this plugin PR. Makes it more easy to track what changed when.

File: docs/user_guide/plugins/treelayercontrol.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+## Simple example
+
+```{code-cell} ipython3
+import folium
+from folium.plugins.treelayercontrol import TreeLayerControl
+from folium.features import Marker
+
+m = folium.Map(location=[46.603354, 1.8883335], zoom_start=5)
+osm = folium.TileLayer("openstreetmap").add_to(m)
```
Comment (Conengmo) [inline docs/user_guide/plugins/treelayercontrol.md:23]: This is not necessary is it?

```suggestion
osm = folium.TileLayer("openstreetmap").add_to(m)
```

**Review by Conengmo — COMMENTED** [review]
Looks good! Useful plugin, nice implementation. Very complete with documentation.

Some very minor comments. I do want to split out the `tojavascript` part in a separate PR, so we'd have to update this branch when that has been merged.

**Comment by Conengmo** [comment]
Update this branch once https://github.com/python-visualization/folium/pull/1912 has been merged.

File: tests/test_template.py (hunk 1 above)
Comment (hansthen) [reply tests/test_template.py:30]: I think the tests you added in a different PR are way more comprehensive. I will rebase this on main.

**Comment by Conengmo** [comment]
For some reason the updating from main didn't go as expected, I'm seeing the changes that are in main also in the diff here. Maybe something to be mindful of next time. I suppose when I merge this PR it will not cause an issue.

**Pull request merged**